criu/apparmor.c

```diff
@@ -0,0 +1,538 @@
+	FILE *f;
+
+	f = fopen(AA_SECURITYFS_PATH "/features/domain/stack", "r");
+	if (!f)
+		return false;
+
+	ret = fscanf(f, "%48s", contents);
+	fclose(f);
+	if (ret != 1) {
+		pr_err("scanning aa stack feature failed\n");
+		return false;
+	}
+
+	if (strcmp("yes", contents)) {
+		pr_warn("aa stack featured disabled: %s\n", contents);
+		return false;
+	}
+
+	f = fopen(AA_SECURITYFS_PATH "/features/domain/version", "r");
+	if (!f)
+		return false;
+
+	ret = fscanf(f, "%d.%d", &major, &minor);
+	fclose(f);
+	if (ret != 2) {
+		pr_err("scanning aa stack version failed\n");
+		return false;
+	}
+
+	return major >= 1 && minor >= 2;
+}
+
+static int restore_aa_namespace(AaNamespace *ns, char *path, int offset)
+{
+	pid_t pid;
+	int status;
+
+	pid = fork();
+	if (pid < 0) {
+		pr_perror("fork failed");
+		return -1;
+	}
+
+	if (!pid) {
+		int i, my_offset, ret, fd;
+		char buf[1024];
+
+		ret = snprintf(buf, sizeof(buf), "changeprofile :%s:", ns->name);
+		if (ret < 0 || ret >= sizeof(buf)) {
+			pr_err("profile %s too big\n", ns->name);
+			exit(1);
+		}
+
+		my_offset = snprintf(path+offset, PATH_MAX-offset, "/namespaces/%s", ns->name);
+		if (my_offset < 0 || my_offset >= PATH_MAX-offset) {
+			pr_err("snprintf'd too many characters\n");
+			exit(1);
+		}
+
+		if (mkdir(path, 0755) < 0) {
+			if (errno == EEXIST) {
+				pr_warn("apparmor namespace %s already exists, restoring into it\n", path);
+			} else {
+				pr_perror("failed to create namespace %s", path);
+				exit(1);
+			}
+		}
+
+		fd = open_proc_rw(PROC_SELF, "attr/current");
+		if (fd < 0) {
+			pr_perror("couldn't open attr/current");
+			goto fail;
+		}
+
+		errno = 0;
+		ret = write(fd, buf, strlen(buf));
+		close(fd);
+		if (ret != strlen(buf)) {
+			pr_perror("failed to change aa namespace %s", buf);
+			goto fail;
+		}
+
+		for (i = 0; i < ns->n_namespaces; i++) {
+			if (restore_aa_namespace(ns, path, offset + my_offset) < 0)
+				goto fail;
+		}
+
+		for (i = 0; i < ns->n_policies; i++) {
+			int fd, n;
+			AaPolicy *p = ns->policies[i];
+
+			fd = open(AA_SECURITYFS_PATH "/.replace", O_WRONLY);
+			if (fd < 0) {
+				pr_perror("couldn't open apparmor load file");
+				goto fail;
+			}
+
+			n = write(fd, p->blob.data, p->blob.len);
+			close(fd);
+			if (n != p->blob.len) {
+				pr_perror("write AA policy failed");
+				goto fail;
+			}
+		}
+
+		exit(0);
+fail:
+		rmdir(path);
+		exit(1);
+	}
+
+	if (waitpid(pid, &status, 0) < 0) {
+		pr_perror("waitpid failed");
+		return -1;
+	}
+
+	if (WIFEXITED(status) && WEXITSTATUS(status) == 0)
+		return 0;
+
+	pr_err("failed to restore aa namespace, worker exited: %d\n", status);
+	return -1;
+}
+
+int prepare_apparmor_namespaces(void)
+{
+	struct cr_img *img;
+	int ret, i;
+	ApparmorEntry *ae;
+
+	img = open_image(CR_FD_APPARMOR, O_RSTR);
+	if (!img)
+		return -1;
+
+	ret = pb_read_one_eof(img, &ae, PB_APPARMOR);
+	close_image(img);
+	if (ret <= 0)
+		return 0; /* there was no AA namespace entry */
+
+	BUG_ON(!ae);
```

Comment:
It would be good if we could reduce the amount of BUG_ON() (https://github.com/checkpoint-restore/criu/issues/767).

Reply:
Ok, removing the BUG_ON.